events/nats.go

```diff
@@ -0,0 +1,377 @@
+		)
+	}
+
+	if err := parameters.validate(); err != nil {
+		return nil, err
+	}
+
+	return &NatsJetstream{parameters: &parameters}, nil
+}
+
+// Open connects to the NATS Jetstream.
+func (n *NatsJetstream) Open() error {
+	if n.parameters == nil {
+		return errors.Wrap(ErrNatsConfig, "NATS config parameters not defined")
+	}
+
+	opts := []nats.Option{
+		nats.Name(n.parameters.AppName),
+		nats.Timeout(n.parameters.ConnectTimeout),
+		nats.RetryOnFailedConnect(true),
+		nats.MaxReconnects(-1),
+		nats.ReconnectWait(reconnectWait),
+		nats.ReconnectJitter(reconnectJitter, reconnectJitter),
+	}
+
+	if n.parameters.StreamUser != "" {
+		opts = append(opts, nats.UserInfo(n.parameters.StreamUser, n.parameters.StreamPass))
+	} else {
+		opts = append(opts, nats.UserCredentials(n.parameters.CredsFile))
+	}
+
+	conn, err := nats.Connect(n.parameters.URL, opts...)
+	if err != nil {
+		return errors.Wrap(ErrNatsConn, err.Error())
+	}
+
+	n.conn = conn
+
+	// setup the channel for subscribers to read messages from.
+	n.subscriberCh = make(MsgCh)
+
+	// setup Jetstream and consumer
+	return n.setup()
+}
+
+func (n *NatsJetstream) setup() error {
+	js, err := n.conn.JetStream()
+	if err != nil {
+		return errors.Wrap(ErrNatsJetstream, err.Error())
+	}
+
+	n.jsctx = js
+
+	if n.parameters.Stream != nil {
+		if err := n.addStream(); err != nil {
+			return err
+		}
+	}
+
+	if n.parameters.Consumer != nil {
+		if err := n.addConsumer(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (n *NatsJetstream) addStream() error {
+	if n.jsctx == nil {
+		return errors.Wrap(ErrNatsJetstreamAddStream, "Jetstream context is not setup")
+	}
+
+	// check stream isn't already present
+	for name := range n.jsctx.StreamNames() {
+		if name == n.parameters.Stream.Name {
+			return nil
+		}
+	}
+
+	var retention nats.RetentionPolicy
+
+	switch n.parameters.Stream.Retention {
+	case "workQueue":
+		retention = nats.WorkQueuePolicy
+	case "limits":
+		retention = nats.LimitsPolicy
+	case "interest":
+		retention = nats.InterestPolicy
+	default:
+		return errors.Wrap(ErrNatsConfig, "unknown retention policy defined: "+n.parameters.Stream.Retention)
+	}
+
+	_, err := n.jsctx.AddStream(
+		&nats.StreamConfig{
+			Name:      n.parameters.Stream.Name,
+			Subjects:  n.parameters.Stream.Subjects,
+			Retention: retention,
+		},
+	)
+
+	if err != nil {
+		return errors.Wrap(ErrNatsJetstreamAddStream, err.Error())
+	}
+
+	return nil
+}
+
+// AddConsumer adds a consumer for a stream
+//
+// Consumers are view into a NATs Jetstream
+// multiple applications may bind to a consumer.
+func (n *NatsJetstream) addConsumer() error {
+	if n.jsctx == nil {
+		return errors.Wrap(ErrNatsJetstreamAddConsumer, "Jetstream context is not setup")
+	}
+
+	// lookup consumers in stream before attempting to add consumer
+	for name := range n.jsctx.ConsumerNames(n.parameters.Stream.Name) {
+		if name == n.parameters.AppName {
+			return nil
+		}
+	}
```

Comment:
Thanks for calling this out, fixed the configuration to require a `consumer` name and use that instead of this ^ magic [98c60b1](https://github.com/metal-toolbox/hollow-toolbox/pull/90/commits/98c60b1f2b7c78a1eeab520f3e21781daa9c60b7)